syncapi/storage/shared/syncserver.go

```diff
@@ -739,13 +741,8 @@ func (d *Database) GetStateDeltas(
 	// handle newly joined rooms and non-joined rooms
 	for roomID, stateStreamEvents := range state {
 		for _, ev := range stateStreamEvents {
-			// TODO: Currently this will incorrectly add rooms which were ALREADY joined but they sent another no-op join event.
-			//       We should be checking if the user was already joined at fromPos and not proceed if so. As a result of this,
-			//       dupe join events will result in the entire room state coming down to the client again. This is added in
-			//       the 'state' part of the response though, so is transparent modulo bandwidth concerns as it is not added to
-			//       the timeline.
 			if membership := getMembershipFromEvent(ev.Event, userID); membership != "" {
-				if membership == gomatrixserverlib.Join {
+				if _, ok := joinedRoomMap[roomID]; !ok && membership == gomatrixserverlib.Join {
```

Comment:
Are we sure that `joinedRoomMap` is getting joined rooms from _before_ the `state` array?

Reply:
Actually, that's a rather good point. No, it probably won't.